scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,256 @@
+        let cp_maker = CDCDefaultCheckpointMaker {
+            session,
+            keyspace,
+            checkpoint_tablename,
+            prepared_stmt,
+        };
+
+        Ok(cp_maker)
+    }
+
+    // Creates new table with specified schema in the given keyspace with provided name.
+    // If a table with such name already exists, doesn't do anything.
+    async fn create_table(
+        session: &Session,
+        keyspace: &str,
+        checkpoint_tablename: &str,
+    ) -> anyhow::Result<()> {
+        let schema = format!(
+            "CREATE TABLE IF NOT EXISTS {}.{} (
+            stream_id blob PRIMARY KEY,
+            time timestamp,
+            generation timestamp,
+            )
+            WITH CLUSTERING ORDER BY (generation DESC)",
+            keyspace, checkpoint_tablename
+        );
+
+        session.query(schema, ()).await?;
+
+        Ok(())
+    }
+}
+
+impl Clone for CDCDefaultCheckpointMaker {
+    fn clone(&self) -> Self {
+        CDCDefaultCheckpointMaker {
+            session: self.session.clone(),
+            keyspace: self.keyspace.clone(),
+            checkpoint_tablename: self.checkpoint_tablename.clone(),
+            prepared_stmt: self.prepared_stmt.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl CDCCheckpointMaker for CDCDefaultCheckpointMaker {
+    /// Writes new record containing given timestamp to the checkpoint table.
+    async fn make_checkpoint(&self, checkpoint: Checkpoint) -> anyhow::Result<()> {
+        let timestamp = Timestamp(chrono::Duration::from_std(checkpoint.0).unwrap());
+
+        self.session
+            .execute(
+                &self.prepared_stmt,
+                (timestamp, &checkpoint.1, &checkpoint.2),
+            )
+            .await?;
+        Ok(())
+    }
+
+    /// Loads last seen generation timestamp or 0 if none.
+    async fn load_last_generation(&self) -> anyhow::Result<Option<GenerationTimestamp>> {
+        let generation = self
+            .session
+            .query(
+                format!(
+                    "SELECT generation FROM {}.{}
+                    LIMIT 1",
+                    self.keyspace, self.checkpoint_tablename
+                ),
+                (),
+            )
+            .await?
+            .maybe_first_row_typed::<GenerationTimestamp>()?;
+
+        Ok(generation)
+    }
+
+    async fn load_last_checkpoint(
+        &self,
+        stream_id: &StreamID,
+    ) -> anyhow::Result<Option<chrono::Duration>> {
+        let timestamp = self
+            .session
+            .query(
+                format!(
+                    "SELECT time FROM {}.{}\
+                    WHERE stream_id = (?)\
+                    LIMIT 1",
+                    self.keyspace, self.checkpoint_tablename
+                ),
+                (stream_id,),
+            )
+            .await?
+            .maybe_first_row_typed::<(chrono::Duration,)>()?
+            .map(|time| -> chrono::Duration { time.0 });
```

Comment:
Why not just `.map(|time| time.0))`?

Privately, I would be tempted to just write

```rust
Ok(self
    .session
    .query(
        format!(
            "SELECT time FROM {}.{}\
            WHERE stream_id = (?)\
            LIMIT 1",
            self.keyspace, self.checkpoint_tablename
        ),
        (stream_id,),
    )
    .await?
    .maybe_first_row_typed::<(chrono::Duration,)>()?
    .map(|time| time.0))
```